core/dbt/contracts/graph/manifest.py

```diff
@@ -1,23 +1,174 @@
+    def original_file_path(self):
+        return os.path.join(self.searched_path, self.relative_path)
+
+
+@dataclass
+class FileHash(JsonSchemaMixin):
+    name: str  # the hash type name
+    checksum: str  # the hashlib.hash_type().hexdigest() of the file contents
+
+    @classmethod
+    def empty(cls):
+        return FileHash(name='none', checksum='')
+
+    @classmethod
+    def path(cls, path: str):
+        return FileHash(name='path', checksum=path)
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return False
+
+        if self.name == 'none' or self.name != other.name:
+            return False
+
+        return self.checksum == other.checksum
+
+    def compare(self, contents: str) -> bool:
+        """Compare the file contents with the given hash"""
+        if self.name == 'none':
+            return False
+
+        return self.from_contents(contents, name=self.name) == self.checksum
+
+    @classmethod
+    def from_contents(cls, contents: str, name='sha256'):
+        """Create a file hash from the given file contents. The hash is always
+        the utf-8 encoding of the contents given, because dbt only reads files
+        as utf-8.
+        """
+        data = contents.encode('utf-8')
+        checksum = hashlib.new(name, data).hexdigest()
+        return cls(name=name, checksum=checksum)
+
+
+@dataclass
+class RemoteFile(JsonSchemaMixin):
+    @property
+    def searched_path(self) -> str:
+        return 'from remote system'
+
+    @property
+    def relative_path(self) -> str:
+        return 'from remote system'
+
+    @property
+    def absolute_path(self) -> str:
+        return 'from remote system'
+
+    @property
+    def original_file_path(self):
+        return 'from remote system'
+
+
+@dataclass
+class SourceFile(JsonSchemaMixin):
+    """Define a source file in dbt"""
+    path: Union[FilePath, RemoteFile]  # the path information
+    checksum: FileHash
+    # we don't want to serialize this
+    _contents: Optional[str] = None
+    # the unique IDs contained in this file
+    nodes: List[str] = field(default_factory=list)
+    docs: List[str] = field(default_factory=list)
+    macros: List[str] = field(default_factory=list)
+    sources: List[str] = field(default_factory=list)
+    # any node patches in this file. The entries are names, not unique ids!
+    patches: List[str] = field(default_factory=list)
+
+    @property
+    def search_key(self) -> Optional[str]:
+        if isinstance(self.path, RemoteFile):
+            return None
+        if self.checksum.name == 'none':
+            return None
+        return self.path.search_key
+
+    @classmethod
+    def from_file(cls, path: FilePath) -> 'SourceFile':
+        file_contents = load_file_contents(path.absolute_path, strip=False)
+        checksum = FileHash.from_contents(file_contents)
+        self = cls(path=path, checksum=checksum)
+        self._contents = file_contents.strip()
+        return self
```

Comment:
this is the only thing i take issue with architecturally in this whole PR. we should keep business logic and I/O separate. introducing a dependency from our pure business logic world (dataclasses) onto our pure I/O world (clients.system) is a bad idea.

IMO this function should take a second arg which is either `file_contents: str` or the more pythonic and possibly more performant `file_handle: typing.TextIO`. then you should push the actual reading of the file into the parser.

in general, i'd prefer the `contracts` namespace to not rely on other namespaces that perform I/O. it makes them more pure, easier to test, and easier to reason about.

Reply:
Sure. I don't really agree about that (I think pushing this into parsing makes it harder to follow what's going on and easier to mess up) but if it's important to you I'm happy to change it.